File: packages/doldskrift-js/src/discovery.ts
import {
  PROTOCOL_VERSION,
  isSymbol,
  parseMode,
  type Detection,
  type DetectionSource,
  type Mode,
} from "./types.js";

function none(): Detection {
  return {
    detected: false,
    version: null,
    mode: null,
    source: null,
    raw: null,
  };
}

/** Parse `content` from `<meta name="doldskrift" content="version=1;mode=encoded">`. */
export function detectFromHtmlMeta(content: string): Detection {
  let version: number | null = null;
  let mode: Mode | null = null;
  for (const part of content.split(";")) {
    const trimmed = part.trim();
    if (trimmed.startsWith("version=")) {
      const v = Number.parseInt(trimmed.slice("version=".length).trim(), 10);
      if (!Number.isNaN(v)) version = v;
    } else if (trimmed.startsWith("mode=")) {
      mode = parseMode(trimmed.slice("mode=".length).trim());
    }
  }
  if (version === null && mode === null && content.trim() === "") {
    return none();
  }
  return {
    detected: true,
    version: version ?? PROTOCOL_VERSION,
    mode,
    source: "html_meta" satisfies DetectionSource,
    raw: content,
  };
}

/** Detect from `data-doldskrift` and optional `data-doldskrift-mode`. */
export function detectFromAttrs(
  versionAttr: string | null | undefined,
  modeAttr: string | null | undefined,
): Detection {
  if (versionAttr == null) return none();
  const parsed = Number.parseInt(versionAttr, 10);
  const version = Number.isNaN(parsed) ? PROTOCOL_VERSION : parsed;
  const mode = modeAttr != null ? parseMode(modeAttr) : null;
  return {
    detected: true,
    version,
    mode,
    source: "data_attributes",
    raw: `data-doldskrift=${versionAttr};data-doldskrift-mode=${modeAttr ?? ""}`,
  };
}

/** Heuristic: fraction of PUA alphabet characters above threshold. */
export function detectPuaHeuristic(text: string, threshold = 0.5): Detection {
  const chars = [...text];
  const total = chars.length;
  if (total === 0) return none();
  let symbols = 0;
  for (const ch of chars) {
    if (isSymbol(ch)) symbols++;
  }
  const ratio = symbols / total;
  if (ratio >= threshold) {
    return {
      detected: true,
      version: PROTOCOL_VERSION,
      mode: "encoded",
      source: "pua_heuristic",
      raw: `ratio=${ratio.toFixed(3)}`,
    };
  }
  return none();
}

export interface DetectInput {
  metaContent?: string;
  dataVersion?: string;
  dataMode?: string;
  text?: string;
}

/**
 * Detect Doldskrift from HTML meta, data attributes, and/or PUA text heuristics.
 * Preference order: meta → data attributes → PUA heuristic.
 */
export function detectDoldskrift(input: DetectInput): Detection {
  if (input.metaContent != null && input.metaContent !== "") {
    const d = detectFromHtmlMeta(input.metaContent);
    if (d.detected) return d;
  }
  if (input.dataVersion != null) {
    const d = detectFromAttrs(input.dataVersion, input.dataMode);
    if (d.detected) return d;
  }
  if (input.text != null) {
    return detectPuaHeuristic(input.text);
  }
  return none();
}

File: crates/doldskrift/src/protocol/discovery.rs
//! Programmatic and HTML discovery helpers.

use crate::{Mode, PROTOCOL_VERSION};
use serde::{Deserialize, Serialize};

/// Where Doldskrift metadata was found.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum DetectionSource {
    /// `<meta name="doldskrift" content="...">`
    HtmlMeta,
    /// `data-doldskrift` / `data-doldskrift-mode` attributes.
    DataAttributes,
    /// Binary magic / container.
    ContainerMagic,
    /// Heuristic: high density of PUA alphabet symbols.
    PuaHeuristic,
}

/// Structured detection result (prefer this over a bare boolean).
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct Detection {
    /// True when Doldskrift was detected.
    pub detected: bool,
    /// Protocol version if known.
    pub version: Option<u8>,
    /// Mode if known.
    pub mode: Option<Mode>,
    /// Detection source.
    pub source: Option<DetectionSource>,
    /// Raw content / attribute payload for debugging.
    pub raw: Option<String>,
}

impl Detection {
    /// Negative result.
    pub fn none() -> Self {
        Self {
            detected: false,
            version: None,
            mode: None,
            source: None,
            raw: None,
        }
    }
}

/// Parse `content` from `<meta name="doldskrift" content="version=1;mode=encoded">`.
pub fn detect_from_html_meta(content: &str) -> Detection {
    let mut version = None;
    let mut mode = None;
    for part in content.split(';') {
        let part = part.trim();
        if let Some(v) = part.strip_prefix("version=") {
            version = v.trim().parse().ok();
        } else if let Some(m) = part.strip_prefix("mode=") {
            mode = m.trim().parse().ok();
        }
    }
    if version.is_none() && mode.is_none() && content.trim().is_empty() {
        return Detection::none();
    }
    Detection {
        detected: true,
        version: version.or(Some(PROTOCOL_VERSION)),
        mode,
        source: Some(DetectionSource::HtmlMeta),
        raw: Some(content.to_owned()),
    }
}

/// Detect from `data-doldskrift` and optional `data-doldskrift-mode`.
pub fn detect_from_attrs(version_attr: Option<&str>, mode_attr: Option<&str>) -> Detection {
    match version_attr {
        None => Detection::none(),
        Some(v) => {
            let version = v.parse().ok().or(Some(PROTOCOL_VERSION));
            let mode = mode_attr.and_then(|m| m.parse().ok());
            Detection {
                detected: true,
                version,
                mode,
                source: Some(DetectionSource::DataAttributes),
                raw: Some(format!(
                    "data-doldskrift={v};data-doldskrift-mode={}",
                    mode_attr.unwrap_or("")
                )),
            }
        }
    }
}

/// Heuristic: fraction of PUA alphabet characters above threshold.
pub fn detect_pua_heuristic(text: &str, threshold: f64) -> Detection {
    let total = text.chars().count();
    if total == 0 {
        return Detection::none();
    }
    let symbols = text.chars().filter(|c| crate::is_symbol(*c)).count();
    let ratio = symbols as f64 / total as f64;
    if ratio >= threshold {
        Detection {
            detected: true,
            version: Some(PROTOCOL_VERSION),
            mode: Some(Mode::Encoded),
            source: Some(DetectionSource::PuaHeuristic),
            raw: Some(format!("ratio={ratio:.3}")),
        }
    } else {
        Detection::none()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn meta_parse() {
        let d = detect_from_html_meta("version=1;mode=encoded");
        assert!(d.detected);
        assert_eq!(d.mode, Some(Mode::Encoded));
    }
}
